src/compiler/checker.ts

```diff
@@ -674,6 +676,12 @@ namespace ts {
             return result;
         }
 
+        function doesNameResolveToLocalSymbolForExportDefault(exports: SymbolTable, meaning: SymbolFlags, name: string): boolean {

```

Comment:
However, it would be simpler to just flip the two tests inside the block that handles external modules. Tests all pass for the code below:

```typescript
    case SyntaxKind.ModuleDeclaration:
        const moduleExports = getSymbolOfNode(location).exports;
        if (location.kind === SyntaxKind.SourceFile ||
            (location.kind === SyntaxKind.ModuleDeclaration && (<ModuleDeclaration>location).name.kind === SyntaxKind.StringLiteral)) {

            // It's an external module. First see if the module has an export default and if the local
            // name of that export default matches.
            if (result = moduleExports["default"]) {
                const localSymbol = getLocalSymbolForExportDefault(result);
                if (localSymbol && (result.flags & meaning) && localSymbol.name === name) {
                    break loop;
                }
                result = undefined;
            }

            // Because of module/namespace merging, a module's exports are in scope, yet we never want to
            // treat an export specifier as putting a member in scope.Therefore, if the name we find is
            // purely an export specifier, it is not actually considered in scope.
            // Two things to note about this:
            //     1. We have to check this without calling getSymbol. The problem with calling getSymbol
            //        on an export specifier is that it might find the export specifier itself, and try to
            //        resolve it as an alias. This will cause the checker to consider the export specifier
            //        a circular alias reference when it might not be.
            //     2. We check === SymbolFlags.Alias in order to check that the symbol is *purely*
            //        an alias. If we used &, we'd be throwing out symbols that have non alias aspects,
            //        which is not the desired behavior.
            if (hasProperty(moduleExports, name) &&
                moduleExports[name].flags === SymbolFlags.Alias &&
                getDeclarationOfKind(moduleExports[name], SyntaxKind.ExportSpecifier)) {
                break;
            }
        }

        if (result = getSymbol(moduleExports, name, meaning & SymbolFlags.ModuleMember)) {
            break loop;
        }
        break;
```